File: undirected_graph.py
class graph:
    adj_list={}
    def __init__(self,vertex):

        self.adj_list[vertex]=[]

    def add_vertex(self, vertex):
        if self.adj_list.get(vertex):
            print('vertex already exists')
        else:
            self.adj_list[vertex]=[]

    def add_edge(self,v1,v2):
        if self.adj_list.get(v1) is not None and self.adj_list.get(v2) is not None:
            if v2 not in self.adj_list[v1] and v1 not in self.adj_list[v2]:
                self.adj_list[v1].append(v2)
                self.adj_list[v2].append(v1)
            else:
                print('Edge already exists')

        else:
            print('one or more vertext don\'t exist in the graph')


    def __str__(self):
        if len(self.adj_list) < 0:
            return 'empty graph'
        else:
            for i in range(len(self.adj_list)):
                l= list(self.adj_list.keys())[i],'==========>',*self.adj_list[list(self.adj_list.keys())[i]]
                print(*l)
            return ''

print("created a new graph with vertex 23")
graph = graph(23)
print(graph)
print('add another vertex 2')
graph.add_vertex(2)
print(graph)
print('add an edge between 23 and 2')
graph.add_edge(23,2)
print(graph)
print('add another vertex 1')
graph.add_vertex(1)
print(graph)
print('add an edge between 1 and 2')
graph.add_edge(1,2)
print(graph)
print('try to add another edge between 23 and 2')
graph.add_edge(23,2)
print(graph)
print('add another vertex 3')
graph.add_vertex(3)
print(graph)
print('add an edge between 23 and 3')
graph.add_edge(23,3)
print(graph)

File: undirected_graph.cpp
#include<iostream>
#include<unordered_map>
#include<algorithm>
#include<vector>

using namespace std;
class graph{

public:
    unordered_map<int, vector<int>> g;
    graph(int x){
    vector<int> j;
    g[x] =j;
    }
    void add_vertex(int x){
    vector <int> j;
    if(g.find(x)==g.end()){
    g[x] = j;
    }
    else{
        cout<<"vertex already exists\n";
        cout<<g.size()<<"\n";
    }
    }
    void add_edge(int v1, int v2){
        if(g.find(v1)!=g.end() && g.find(v2)!=g.end()){
        if(find(g[v1].begin(),g[v1].end(),v2)==g[v1].end()){
        g[v1].push_back(v2);
        }
        if(find(g[v2].begin(),g[v2].end(),v1)==g[v2].end()){

        g[v2].push_back(v1);}
    }
    else{
        cout<<"One or both vertex do not exist\n";
    }
    }
};

int main(){

graph gg(2);
gg.add_vertex(3);
gg.add_edge(2,3);
gg.add_vertex(4);
gg.add_edge(2,4);
gg.add_vertex(6);
gg.add_edge(4,6);



for(auto x: gg.g){
        cout<<x.first<<"->";
        for(int i = 0; i < x.second.size(); i++){
            cout<<" "<<x.second[i];
        }
        cout<<"\n";
}
}

